src/vm/jitinterface.cpp

```diff
@@ -11815,6 +11815,79 @@ void* CEEJitInfo::getFieldAddress(CORINFO_FIELD_HANDLE fieldHnd,
     return result;
 }
 
+/*********************************************************************/
+CORINFO_CLASS_HANDLE CEEJitInfo::getStaticFieldCurrentClass(CORINFO_FIELD_HANDLE fieldHnd,
+                                                            bool* isInitOnly)
+{
+    CONTRACTL {
+        SO_TOLERANT;
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    CORINFO_CLASS_HANDLE result = NULL;
+
+    if (isInitOnly != NULL)
+    {
+        *isInitOnly = false;
+    }
+
+    // Only examine the field's value if we are producing jitted code.
+    if (isVerifyOnly() || IsCompilingForNGen() || IsReadyToRunCompilation())
+    {
+        return result;
+    }
+
+    JIT_TO_EE_TRANSITION();
+
+    FieldDesc* field = (FieldDesc*) fieldHnd;
+
+    // We're only interested in ref class typed static fields
+    // where the field handle specifies a unique location.
+    if (field->IsStatic() && field->IsObjRef() && !field->IsThreadStatic())
+    {
+        MethodTable* pEnclosingMT = field->GetEnclosingMethodTable();
+
+        // We must not call here for statics of collectible types.
+        _ASSERTE(!pEnclosingMT->Collectible());
+
+        if (!pEnclosingMT->IsSharedByGenericInstantiations())
+        {
+            // Allocate space for the local class if necessary, but don't trigger
+            // class construction.
+            DomainLocalModule *pLocalModule = pEnclosingMT->GetDomainLocalModule();
+            pLocalModule->PopulateClass(pEnclosingMT);
+            
+            GCX_COOP();
+            
+            OBJECTREF fieldObj = field->GetStaticOBJECTREF();
+            VALIDATEOBJECTREF(fieldObj);
+
+            if (fieldObj != NULL)
+            {
+                MethodTable *pObjMT = fieldObj->GetMethodTable();
+                
+                // TODO: Check if the jit is allowed to embed this handle in jitted code.
+                // Note for the initonly cases it probably won't embed.
+                result = (CORINFO_CLASS_HANDLE) pObjMT;
+            }
+        }
+    }
+    
+    // If the field is init only, and the class is known, the class can't ever change.
```

Comment:
This is not sufficient condition. The static constructor has to be finished running for this to be true. Consider this:

```
using System;

class Test
{
    static readonly object s;

    static Test()
    {
       s = new object();
       DoStuff();
       s = "Method";
    }

    static void DoStuff()
    {
        Console.WriteLine(s.GetType());
    }

    static void Main()
    {
    }
}
```

Reply:
Good point -- is checking that `pEnclosingMT->IsClassInited()` sufficient?

Reply:
Yes, it should be. You may want to do it inside the block above only when you are returning the classhandle (e.g. IsClassInited does not work when pEnclosingMT is shared generic).

Reply:
Also, `isInitOnly` is not the best name of the flag once you fix this...

Reply:
Suggestions?

Reply:
`bool *pIsFinal` or `bool *pIsImmutable`?

Or flip this around and have `bool *pIsSpeculative` flag instead. When the EE returns `*pIsSpeculative = true`, it would mean that the answer is a guess that the JIT has to emit a test for. I think `bool *pIsSpeculative` would work well as a general pattern for other APIs (e.g. devirtualization). If the JIT is not interested in speculative answer, it would pass `pIsSpeculative: NULL`.